### PR TITLE
Full refactor of the vastClient logic to get the vastResponse to better track errors on VASTWaterfall

### PR DESCRIPTION
**Pull request by carpasse**

### CODE REVIEW

File: src/ads/vast/client/VASTClient.js [hunk 1]
```diff
@@ -208,25 +156,65 @@ VASTClient.prototype._buildAd = function buildAd(adJxonTree) {
...
       if (inLine.creatives.length === 0) {
-        throw new VASTError("on VASTClient._buildAd, missing creative in InLine element", 101);
+        return new VASTError("on VASTClient.getVASTAd.validateAd, missing creative in InLine element", 101);
       }
     }
 
     if (wrapper) {
       if (!wrapper.VASTAdTagURI) {
-        throw new VASTError("on VASTClient._buildAd, missing 'VASTAdTagURI' in wrapper", 101);
+        return new VASTError("on VASTClient.getVASTAd.validateAd, missing 'VASTAdTagURI' in wrapper", 101);
```
Comment (Fetz) [inline src/ads/vast/client/VASTClient.js]: A lot of repetition  with "on VASTClient.getVASTAd.validateAd"

File: src/ads/vast/client/VASTClient.js [hunk 2]
```diff
@@ -14,14 +14,20 @@ function VASTClient(options) {
 VASTClient.prototype.getVASTResponse = function getVASTResponse(adTagUrl, callback) {
   var that = this;
 
-  //We reset the errorURLMacros before doing anything.
-  this.errorURLMacros = [];
+  var error = sanityCheck(adTagUrl, callback);
+  if (error) {
```
Comment (Fetz) [inline src/ads/vast/client/VASTClient.js]: I would move this in to a function

File: src/ads/vast/client/VASTClient.js [hunk 3]
```diff
@@ -14,14 +14,20 @@ function VASTClient(options) {
...
+    if (isFunction(callback)) {
+      callback(error);
```
Comment (Fetz) [inline src/ads/vast/client/VASTClient.js]: Even if there is an error you want still to allow the code to run?

File: src/ads/vast/client/VASTClient.js (hunk 1 above)
Comment (carpasse) [inline src/ads/vast/client/VASTClient.js]: The Idea is to be able to know where the error happened by just looking at the error message.
How would your rewrite it?

File: src/ads/vast/client/VASTClient.js (hunk 3 above)
Comment (carpasse) [inline src/ads/vast/client/VASTClient.js]: nop. Thats an error. Writing a fix for it

File: src/ads/vast/client/VASTClient.js [hunk 4]
```diff
@@ -32,174 +38,116 @@ VASTClient.prototype.getVASTResponse = function getVASTResponse(adTagUrl, callba
...
+    getAd(waterfallAds.shift(), [], waterfallHandler);
 
-    if (opts.ads.length >= that.WRAPPER_LIMIT) {
-      return new VASTError("on VASTClient._getAd, players wrapper limit reached (the limit is " + that.WRAPPER_LIMIT + ")", 302);
+    /*** Local functions ***/
+    function waterfallHandler(error, adChain) {
+      if (error) {
+        that._trackError(error, adChain);
+        if (waterfallAds.length > 0) {
+          getAd(waterfallAds.shift(),[], waterfallHandler);
```
Comment (Fetz) [inline src/ads/vast/client/VASTClient.js]: after I would call return;

so you can remove the else and have only callback(error, adChain)

**Comment by Fetz** [comment]
:+1:

**Pull request merged**